pkg/bindings/containers/containers.go

```diff
@@ -333,3 +339,125 @@ func ContainerInit(ctx context.Context, nameOrID string) error {
 	}
 	return response.Process(nil)
 }
+
+// Attach attaches to a running container
+func Attach(ctx context.Context, nameOrId string, detachKeys *string, logs, stream *bool, stdin, stdout, stderr chan string) error {
+	conn, err := bindings.GetClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	params := url.Values{}
+	if detachKeys != nil {
+		params.Add("detachKeys", *detachKeys)
+	}
+	if logs != nil {
+		params.Add("logs", fmt.Sprintf("%t", *logs))
+	}
+	if stream != nil {
+		params.Add("stream", fmt.Sprintf("%t", *stream))
+	}
+	if stdin != nil {
+		params.Add("stdin", "true")
+	}
+	if stdout != nil {
+		params.Add("stdout", "true")
+	}
+	if stderr != nil {
+		params.Add("stderr", "true")
+	}
+
+	response, err := conn.DoRequest(nil, http.MethodPost, "/containers/%s/attach", params, nameOrId)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+
+	ctype := response.Header.Get("Content-Type")
+	upgrade := response.Header.Get("Connection")
+
+	buffer := make([]byte, 1024)
+	if ctype == "application/vnd.docker.raw-stream" && upgrade == "Upgrade" {
+		for {
+			// Read multiplexed channels and write to appropriate stream
+			fd, l, err := DemuxHeader(response.Body, buffer)
+			if err != nil {
+				switch {
+				case errors.Is(err, io.EOF):
+					return nil
+				case errors.Is(err, io.ErrUnexpectedEOF):
+					continue
+				}
+				return err
+			}
+			frame, err := DemuxFrame(response.Body, buffer, l)
+			if err != nil {
+				return nil
+			}
+
+			switch {
+			case fd == 0 && stdin != nil:
+				stdout <- string(frame)
+			case fd == 1 && stdout != nil:
+				stdout <- string(frame)
+			case fd == 2 && stderr != nil:
+				stderr <- string(frame)
+			case fd == 3:
+				return fmt.Errorf("error from daemon in stream: %s", frame)
+			default:
+				return fmt.Errorf("unrecognized input header: %d", fd)
+			}
+		}
+	} else {
+		// If not multiplex'ed from server just dump stream to stdout
+		for {
+			_, err := response.Body.Read(buffer)
+			if err != nil {
+				if !errors.Is(err, io.EOF) {
+					return err
+				}
+				break
+			}
+			stdout <- string(buffer)
+		}
+	}
+	return err
+}
+
+// DemuxHeader reads header for stream from server multiplexed stdin/stdout/stderr/2nd error channel
+func DemuxHeader(r io.Reader, buffer []byte) (fd, sz int, err error) {
```

Comment:
Are we sure this works with terminal attach? There'll be no headers there

Reply:
Do you still send the application/vnd.docker.raw-stream header?

Reply:
Yes - that is always sent, for both types of attach.